### PR TITLE
allow passing additional arguments to scorers in dataops

### PR DESCRIPTION
**Pull request by jeromedockes**
We can now call `.skb.with_scoring` to define which scorer should be used and pass additional arguments to it (such as sample weights). those can be dataops (ie sample weights are computed dynamically from the inputs)

```
>>> import skrub
>>> from sklearn.dummy import DummyClassifier

>>> df = skrub.datasets.toy_products()
>>> data = skrub.var("df", df)
>>> X = data[["description", "price"]].skb.mark_as_X(cv=2)
>>> y = data["category"].skb.mark_as_y()
>>> sample_weight = X["price"]

>>> pred = X.skb.apply(DummyClassifier(), y=y).skb.with_scoring(
...     "accuracy", kwargs={"sample_weight": sample_weight}
... )

>>> pred.skb.cross_validate()
   fit_time  score_time  test_score
0  0.002466    0.003201    0.888889
1  0.002033    0.002878    0.647059
```

### CODE REVIEW

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst [hunk 1]
```diff
@@ -193,3 +193,74 @@ The test set only contains data from the "bestproducts.com" seller.
...
+==========================================
+
+Sometimes we have additional information to pass to the scorer such as sample
+weights, group information etc.
+
+We can control how scoring is performed by using
+:meth:`DataOp.skb.with_scoring`. It has a ``scoring`` parameter, which can be
+anything :func:`sklearn.model_selection.cross_validate` accepts for ``scoring``
+such as a metric name, callable scorer, or dict mapping metric names to scorers
+(see the reference documentation of :meth:`DataOp.skb.with_scoring` for
```
Comment (rcap107) [inline doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: crossreference seems to be broken here

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst [hunk 2]
```diff
@@ -193,3 +193,74 @@ The test set only contains data from the "bestproducts.com" seller.
...
+
+Passing additional arguments to the scorer
+==========================================
+
+Sometimes we have additional information to pass to the scorer such as sample
+weights, group information etc.
+
+We can control how scoring is performed by using
+:meth:`DataOp.skb.with_scoring`. It has a ``scoring`` parameter, which can be
+anything :func:`sklearn.model_selection.cross_validate` accepts for ``scoring``
```
Comment (rcap107) [inline doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: not limited to this PR, I was wondering if here (and elsewhere) we should use the `~` to have the short name of the module rather than linking all of `sklearn.model_selection.cross_validate`

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst [hunk 3]
```diff
@@ -193,3 +193,74 @@ The test set only contains data from the "bestproducts.com" seller.
...
+
+>>> sample_weight = X["price"]
+>>> pred.skb.with_scoring(
+...     "accuracy", kwargs={"sample_weight": sample_weight}
+... ).skb.cross_validate() # doctest: +SKIP
+   fit_time  score_time  test_accuracy
+0  0.003045    0.003275       0.888889
+1  0.002659    0.003026       0.647059
+
+Besides passing extra arguments, :meth:`DataOp.skb.with_scoring` can also be
```
Comment (rcap107) [inline doc/modules/data_ops/validation/tuning_validating_data_ops.rst:241]: also broken here

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst [hunk 4]
```diff
@@ -193,3 +193,74 @@ The test set only contains data from the "bestproducts.com" seller.
...
+-0.6365141682948128
+
+(The score above is the negative log loss, not the default accuracy which would
+be positive.)
+
+Note that :meth:`DataOp.skb.with_scoring` only changes how scoring is performed
+(the outputs of :meth:`DataOp.skb.cross_validate`,
+:meth:`DataOp.skb.make_randomized_search`, :meth:`SkrubLearner.score` etc.),
+**not** the actual outputs of the learner (it does _not_ affect the outputs of
+:meth:`DataOp.skb.eval`, :meth:`SkrubLearner.predict`, etc.)
```
Comment (rcap107) [inline doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: `SkrubLearner.score` and `SkrubLearner.predict` xrefs are also broken

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst [hunk 5]
```diff
@@ -193,3 +193,74 @@ The test set only contains data from the "bestproducts.com" seller.
...
+>>> learner.fit(split['train'])
+SkrubLearner(data_op=<Scoring <Apply DummyClassifier> (1 scorers)>
+    This DataOp will be scored with:
+      - 'neg_log_loss'
+    Use .skb.cross_validate(…) or .skb.make_learner(…).score(…) to compute scores.)
+>>> learner.score(split['test']) # doctest: +SKIP
+-0.6365141682948128
+
+(The score above is the negative log loss, not the default accuracy which would
+be positive.)
```
Comment (rcap107) [inline doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: Maybe a  rewrite to make it more evident that we're not using the default scoring metric? 
```suggestion
(The score above is `negative` because it is the `negative log loss`, rather than the default accuracy which would
be positive.)
```

File: skrub/_data_ops/tests/test_estimators.py [hunk 6]
```diff
@@ -173,6 +181,152 @@ def test_cross_validate(data_op, data, n_jobs):
     assert score.mean() == pytest.approx(0.84, abs=0.05)
 
 
+def test_cross_validate_with_scoring():
+    X_a, y_a = make_classification(random_state=0)
+    X = skrub.X(X_a)
+
+    def get_sw(X):
```
Comment (rcap107) [inline skrub/_data_ops/tests/test_estimators.py]: I was wondering for a while what "sw" meant here
```suggestion
    def get_sample_weights(X):
```

File: skrub/_data_ops/tests/test_interactive_features.py [hunk 7]
```diff
@@ -276,6 +289,32 @@ def test_repr():
     >>> print(e.skb.describe_param_grid())
     - choose_float(10, 100): choose_float(10, 100)
       choose_float(1, 100, log=True, n_...): choose_float(1, 100, log=True, n_steps=100, default=10)
+    >>> from sklearn.dummy import DummyClassifier
```
Comment (rcap107) [inline skrub/_data_ops/tests/test_interactive_features.py:292]: Unrelated from the current PR, asking here because I just noticed this test. Is this needed because the regular doctests are not sufficient?

File: skrub/_data_ops/_data_ops.py [hunk 8]
```diff
@@ -699,7 +706,7 @@ def __repr__(self):
         if preview is NULL:
             return result
         subsample_msg = " (on a subsample)" if uses_subsampling(self) else ""
-        header = f"Result{subsample_msg}:"
+        header = f"{self._skrub_impl.__skrub_preview_heading__()}{subsample_msg}:"
```
Comment (rcap107) [inline skrub/_data_ops/_data_ops.py:709]: Was the preview_heading function added for possible changes in the future? It doesn't look like it's simplifying things for now

File: skrub/_data_ops/_estimator.py [hunk 9]
```diff
@@ -245,11 +259,22 @@ def report(self, *, environment, mode, **full_report_kwargs):
...
+    def _score(self, environment):
+        score_node = _find_scoring_node(self.data_op)
+        if score_node is None:
+            return self._eval_in_mode("score", environment)
+        estimator = self.__skrub_to_Xy_pipeline__(environment)
+        cv_data = _compute_X_y_and_cv(self.data_op, environment)
+        X, y = cv_data["X"], cv_data.get("y")
+        return estimator._score(X, y, cast_to_float=False)
+
     def __getattr__(self, name):
```
Comment (rcap107) [inline skrub/_data_ops/_estimator.py:271]: Following along the code:

This function is used when the learner is called with predict/predict_proba/score etc. to trigger the evaluation of all the steps in the given mode 

Then return the result of the evaluated learner

If it's asking for the scorer, then more stuff needs to be done which is why it's calling _score instead

File: skrub/_data_ops/_estimator.py [hunk 10]
```diff
@@ -566,6 +601,69 @@ def _eval_in_mode(self, mode, X, y=None):
...
+        scorer = check_scoring(self, scoring)
+        kwargs = kwargs or {}
+        if not hasattr(scorer, "get_metadata_routing"):
+            return partial(scorer, **kwargs)
+        scorer = copy.deepcopy(scorer)
+        if hasattr(scorer, "_kwargs"):
+            scorer._kwargs = {**scorer._kwargs, **kwargs}
+        elif hasattr(scorer, "_scorers"):
+            scorer._scorers = {
+                k: self._prepare_scorer(v, kwargs) for k, v in scorer._scorers.items()
```
Comment (rcap107) [inline skrub/_data_ops/_estimator.py:623]: The recursive call here is to evaluate all the possible nested estimators in case they are dicts that contain dicts, correct? 

The loop ends when the code reaches a scorer, which doesn't have a _scorers attribute (meaning it's not nested anymore), or when it does have _kwargs, in which case the kwargs are set and the scorer is returned

File: skrub/_data_ops/_estimator.py [hunk 11]
```diff
@@ -566,6 +601,69 @@ def _eval_in_mode(self, mode, X, y=None):
...
+        score_node = _find_scoring_node(self.data_op)
+        if score_node is None:
+            return self._eval_in_mode("score", X, y)
+        env = self._get_env(X, y)
+        scorers = evaluate(
+            score_node._skrub_impl.scorers, mode="fit_transform", environment=env
+        )
+        all_scores = []
+        for scorer_info in scorers:
+            scorer = self._prepare_scorer(scorer_info["scoring"], scorer_info["kwargs"])
```
Comment (rcap107) [inline skrub/_data_ops/_estimator.py:656]: I lost track of where is the code that lets me specify a custom name for the scorer

File: skrub/_data_ops/_skrub_namespace.py [hunk 12]
```diff
@@ -2833,6 +2834,141 @@ def is_y(self):
...
+
+        This records a scikit-learn
+        `scorer <https://scikit-learn.org/stable/modules/model_evaluation.html#the-scoring-parameter-defining-model-evaluation-rules>`_
+        to be used when measuring the quality of predictions.
+
+        It allows configuring the default scoring metrics, and most importantly
+        to compute any additional arguments needed for scoring (such as sample
+        weights) as part of the DataOp evaluation.
+
+        This method can be called several times to add multiple scorers that
```
Comment (rcap107) [inline skrub/_data_ops/_skrub_namespace.py]: maybe "chained" to make it clearer that all the calls should come one after the other ?

File: skrub/_data_ops/_skrub_namespace.py [hunk 13]
```diff
@@ -2833,6 +2834,141 @@ def is_y(self):
...
+        use different kwargs. Several scoring strategies can be passed to a
+        single calls if they do not need different kwargs (see the description
+        of the `scoring` parameter).
```
Comment (rcap107) [inline skrub/_data_ops/_skrub_namespace.py]: here and in the other places, but I saw you didn't put the double backticks anywhere
was it intentional?
```suggestion
        of the ``scoring`` parameter).
```

File: skrub/_data_ops/_skrub_namespace.py [hunk 14]
```diff
@@ -2833,6 +2834,141 @@ def is_y(self):
         """Whether this DataOp has been marked with :meth:`.skb.mark_as_y()`."""
         return self._data_op._skrub_impl.is_y
 
+    @checked_data_op_constructor
+    def with_scoring(self, scoring, kwargs=None, name=None):
```
Comment (rcap107) [inline skrub/_data_ops/_skrub_namespace.py:2838]: this wasn't added to the api reference so it doesn't appear in the docs

File: skrub/_data_ops/_skrub_namespace.py [hunk 15]
```diff
@@ -2833,6 +2834,141 @@ def is_y(self):
...
+            with scoring information attached to it, which will be used for
+            example in cross-validation.
+
+        Notes
+        -----
+        If this method is used several times, all calls to it must be grouped
+        -- there can be no other nodes in-between. For example
+        `pred.skb.score_with('accuracy').skb.score_with('roc_auc')` is allowed
+        but not
+        `pred.skb.score_with('accuracy').skb.apply_func(a_function).skb.score_with('roc_auc')`.
```
Comment (rcap107) [inline skrub/_data_ops/_skrub_namespace.py]:
```suggestion
        `pred.skb.score_with('accuracy').skb.score_with('roc_auc')` is allowed,
        while
        `pred.skb.score_with('accuracy').skb.apply_func(a_function).skb.score_with('roc_auc')`
        is not.
```

File: skrub/_data_ops/_skrub_namespace.py [hunk 16]
```diff
@@ -2833,6 +2834,141 @@ def is_y(self):
...
+        >>> pred.skb.with_scoring("accuracy").skb.with_scoring(
+        ...     "accuracy",
+        ...     kwargs={"sample_weight": sample_weight},
+        ...     name="weighted_accuracy",
+        ... ).skb.cross_validate() # doctest: +SKIP
+           fit_time  score_time  test_accuracy  test_weighted_accuracy
+        0  0.002738    0.005733       0.666667                0.888889
+        1  0.002845    0.005705       0.666667                0.647059
+
+        When we have several scorers but they use the same kwargs we can pass a list
```
Comment (rcap107) [inline skrub/_data_ops/_skrub_namespace.py]:
```suggestion
        When we have several scorers that use the same kwargs, we can pass a list
```

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst (hunk 1 above)
Comment (rcap107) [reply doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: I think it's broken because with_scoring was not added to the api reference

File: CHANGES.rst [hunk 17]
```diff
@@ -11,6 +11,11 @@ Ongoing Development
 
 New Features
 ------------
+- It is now possible to pass additional (dynamically computed) arguments to the
+  scorers used by :class:`DataOp` objects for validation, hyperparameter search
+  etc. For example, sample weights. This is achieved by passing the scorers and
+  their arguments to :meth:`DataOp.skb.with_scoring`. :pr:`1995` by
```
Comment (rcap107) [inline CHANGES.rst:17]: xref broken here, should be fixed once the api template is updated

File: skrub/_data_ops/tests/test_interactive_features.py (hunk 7 above)
Comment (jeromedockes) [reply skrub/_data_ops/tests/test_interactive_features.py:292]: yeah it is here to check specific aspects of the object's repr that are not necessarily covered by the main doctests

File: skrub/_data_ops/_data_ops.py (hunk 8 above)
Comment (jeromedockes) [reply skrub/_data_ops/_data_ops.py:709]: it is to allow the node type to override the default which is just "Results". the Scoring node uses it to provide more detail

File: skrub/_data_ops/_estimator.py (hunk 9 above)
Comment (jeromedockes) [reply skrub/_data_ops/_estimator.py:271]: yes 👍

File: skrub/_data_ops/_estimator.py (hunk 10 above)
Comment (jeromedockes) [reply skrub/_data_ops/_estimator.py:623]: yes there will be at most one level of recursion, for MultiMetric scorers

File: skrub/_data_ops/_estimator.py (hunk 11 above)
Comment (jeromedockes) [reply skrub/_data_ops/_estimator.py:656]: `.skb.with_scoring(..., name='the name')`

File: doc/modules/data_ops/validation/tuning_validating_data_ops.rst (hunk 2 above)
Comment (jeromedockes) [reply doc/modules/data_ops/validation/tuning_validating_data_ops.rst]: ok I did it here, we can edit other places in other pr

**Comment by jeromedockes** [comment]
thanks for the review @rcap107 , I checked the links and they should be good now

**Comment by jeromedockes** [comment]
thanks for the reviews :)
